### PR TITLE
2281 - target manually partitioned disk

### PR DESCRIPTION
**Pull request by jimmykarily**
Fixes https://github.com/kairos-io/kairos/issues/2281

### CODE REVIEW

**Comment by jimmykarily** [comment]
This is just playing around with code to find where we break it.

I created a VM in qemu with 2 disks:

    /dev/vda -> 80G
    /dev/vdb -> 40G

Kairos would automatically pick /dev/vda because it's the bigger disk. My goal is to point the installation to a manually partitioned /dev/vdb.

With the (not-to-be-merged) changes in this PR and a config like the one below, I can successfully install on `/dev/sdb`

```
#cloud-config

strict: true
debug: true

install:
  no-format: true
  auto: true
  poweroff: false
  reboot: false
  grub_options:
    extra_cmdline: "rd.immucore.debug"
users:
  - name: "kairos"
    passwd: "kairos"

stages:
  kairos-install.pre.before:
  - if:  '[ -e "/dev/vdb" ]'
    name: "Create partitions"
    commands:
      - |
        parted --script --machine -- "/dev/vdb" mklabel gpt
        parted --script "/dev/vdb" mkpart primary Data 0 1MB
    layout:
      add_partitions:
        - fsLabel: COS_OEM
          size: 64
          pLabel: oem
        - fsLabel: COS_RECOVERY
          size: 8500
          pLabel: recovery
        - fsLabel: COS_STATE
          size: 18000
          pLabel: state
        - fsLabel: COS_PERSISTENT
          pLabel: persistent
          size: 0
          filesystem: "ext4"
  boot:
    - systemd_firstboot:
      keymap: us
```
I manually create a partition of 1Mb to allow grub to install. Putting it to the `add_partitions` key will also format it as `ext2` (and possibly other default) which make grub complain.

File: internal/agent/install.go [hunk 1]
```diff
@@ -213,9 +213,10 @@ func RunInstall(c *config.Config) error {
 	utils.SetEnv(c.Env)
 	utils.SetEnv(c.Install.Env)
 
-	if c.Install.Device == "" || c.Install.Device == "auto" {
-		c.Install.Device = detectDevice()
-	}
+	// if c.Install.Device == "" || c.Install.Device == "auto" {
+	// 	c.Install.Device = detectDevice()
+	// }
+	c.Install.Device = "/dev/vdb"
```
Comment (jimmykarily) [inline internal/agent/install.go:219]: we need to respect `NoFormat: true` and detect the target device using labels. `detectDevice()` will simply point to `/dev/vda` (in my example in the PR description) which is not correct.

We should find the device that has the needed labels, validate that all needed partitions are there and if true, make that the target device.

File: internal/agent/install.go [hunk 2]
```diff
@@ -267,7 +268,9 @@ func runInstall(c *config.Config) error {
 	}
 
 	// TODO: This should not be neccessary
-	installSpec.NoFormat = c.Install.NoFormat
+	fmt.Printf("!!!!!!!!!! installSpec.NoFormat = %+v\n", installSpec.NoFormat)
+	fmt.Printf("!!!!!!!!!! c.Install.NoFormat = %+v\n", c.Install.NoFormat)
+	//installSpec.NoFormat = c.Install.NoFormat
```
Comment (jimmykarily) [inline internal/agent/install.go:273]: This line sets `NoFormat` to `false` although in my config I set it to `true`

**Comment by jimmykarily** [comment]
this is an alternative command for the grub partition:

```
        sgdisk --new=1:2048:+1M --change-name=1:'bios' --typecode=1:EF02 /dev/vdb # for grub
```

File: internal/agent/install.go (hunk 1 above)
Comment (jimmykarily) [reply internal/agent/install.go]: Now there is `DetectPreConfiguredDevice` which looks for a disk with the appropriate labels.

**Comment by jimmykarily** [comment]
The binary from this PR makes the added test pass: https://github.com/kairos-io/kairos/pull/2291

File: config.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+#cloud-config
```
Comment (jimmykarily) [inline config.yaml:1]: This sample is now part of the test in kairos: https://github.com/kairos-io/kairos/pull/2291/files#diff-1ff1699e612ac7f8c508e5f9f6a784b37441b01b8cfdebd8da3b068280385247R87

I'll remove it from here.

File: internal/agent/install.go [hunk 4]
```diff
@@ -213,10 +213,6 @@ func RunInstall(c *config.Config) error {
 	utils.SetEnv(c.Env)
 	utils.SetEnv(c.Install.Env)
 
-	if c.Install.Device == "" || c.Install.Device == "auto" {
-		c.Install.Device = detectDevice()
```
Comment (jimmykarily) [inline internal/agent/install.go:217]: moved here: https://github.com/kairos-io/kairos-agent/pull/235/files#diff-53b724197bc14c3828ce6725322b892cb788860b0526184b63d7d114edb60941R571

File: pkg/uki/install.go [hunk 5]
```diff
@@ -40,17 +40,32 @@ func (i *InstallAction) Run() (err error) {
...
-	if err != nil {
-		i.cfg.Logger.Errorf("deactivating devices: %s", err.Error())
-		return err
-	}
-	// Partition device
-	err = e.PartitionAndFormatDevice(i.spec)
-	if err != nil {
-		i.cfg.Logger.Errorf("partitioning and formating devices: %s", err.Error())
-		return err
+	if i.spec.NoFormat {
```
Comment (jimmykarily) [inline pkg/uki/install.go:43]: I'm trying to add a test for the UKI case too here: https://github.com/kairos-io/kairos/pull/2291
to be complete.

Comment (jimmykarily) [reply pkg/uki/install.go:43]: Manually creating the disk layout that the UKI installation creates is rather tricky:

```
[kairos@fedora ~]$ sudo blkid
/dev/sr0: BLOCK_SIZE="2048" UUID="2024-04-09-09-34-22-00" LABEL="UKI_ISO_INSTALL" TYPE="iso9660"
/dev/loop0: UUID="B9E2-794E" BLOCK_SIZE="512" TYPE="vfat"
/dev/vda2: UUID="8bfa06f9-ca4f-56dc-90c9-49cf20f4f45e" TYPE="crypto_LUKS" PARTLABEL="oem" PARTUUID="d0be3cb8-e04d-4bc9-9bde-311a516eece2"
/dev/vda3: UUID="85c39d0f-4867-5227-8334-f5eec606d9eb" TYPE="crypto_LUKS" PARTLABEL="persistent" PARTUUID="bc2d3f7d-54c6-4fc9-9918-595b19129d50"
/dev/vda1: LABEL_FATBOOT="COS_GRUB" LABEL="COS_GRUB" UUID="E9F3-549E" BLOCK_SIZE="512" TYPE="vfat" PARTLABEL="efi" PARTUUID="feb38d63-49d4-4b2d-a1d6-2f340a27e4c5"
```
We can keep the code around here but writing a test for it is rather complex (creating a layout with encrypted partitions using the tpm). I'm not sure if anyone will ever be interested in this feature in UKI mode.

**Review by Itxaka — APPROVED** [review]
code looks good. Im a bit worried about the auto select and no format stuff, but if it passes tests then its good to me. cant break no-format anymore :D

**Comment by jimmykarily** [comment]
> code looks good. Im a bit worried about the auto select and no format stuff, but if it passes tests then its good to me. cant break no-format anymore :D

I didn't run the full test suite, only the new test. The only thing I can say is that it fixes `no-format` :D . 
Let's try this: https://github.com/kairos-io/kairos/pull/2291/files#diff-2705f6016a64189abd14041e2f3e8dbb829553295d9088e4eb42eccb5e61bf47R299

**Comment by jimmykarily** [comment]
Depends on this: https://github.com/mudler/yip/pull/141

otherwise this config won't work:

```
#cloud-config

strict: true
debug: true

install:
  no-format: true
  auto: false
  poweroff: false
  reboot: false
  grub_options:
    extra_cmdline: "rd.immucore.debug"

users:
  - name: "kairos"
    passwd: "kairos"

stages:
  kairos-install.pre.before:
  - if:  '[ -e "/dev/vdb" ]'
    name: "Create partitions"
    commands:
      - |
        parted --script --machine -- "/dev/vdb" mklabel gpt
        #sgdisk --new=1:2048:+1M --change-name=1:'bios' --typecode=1:EF02 /dev/vdb # for grub
    layout:
      device:
        path: "/dev/vdb"
      add_partitions:
        - fsLabel: COS_GRUB
          size: 64
          pLabel: efi
          filesystem: "fat"
        - fsLabel: COS_OEM
          size: 64
          pLabel: oem
        - fsLabel: COS_RECOVERY
          size: 8500
          pLabel: recovery
        - fsLabel: COS_STATE
          size: 18000
          pLabel: state
        - fsLabel: COS_PERSISTENT
          pLabel: persistent
          size: 0
          filesystem: "ext4"
```

**Comment by jimmykarily** [comment]
> > code looks good. Im a bit worried about the auto select and no format stuff, but if it passes tests then its good to me. cant break no-format anymore :D
> 
> I didn't run the full test suite, only the new test. The only thing I can say is that it fixes `no-format` :D . Let's try this: https://github.com/kairos-io/kairos/pull/2291/files#diff-2705f6016a64189abd14041e2f3e8dbb829553295d9088e4eb42eccb5e61bf47R299

The rest of the tests pass: https://github.com/kairos-io/kairos/actions/runs/8616215034/job/23646545504?pr=2291

but I pushed some fixes here to fix the new test.

**Review by Itxaka — APPROVED** [review]
still looking good to me

**Pull request merged**